### PR TITLE
[PAC][CodeGen][ELF][AArch64] Support signed GOT

### LINKED ISSUE #83801: [PAC][ELF] Add codegen support for signed GOT

### PR DESCRIPTION
**Pull request by kovdan01**
Depends on #96158 and #96159

Support the following relocations and assembly operators:

- `R_AARCH64_AUTH_ADR_GOT_PAGE` (`:got_auth:` for `adrp`)
- `R_AARCH64_AUTH_LD64_GOT_LO12_NC` (`:got_auth_lo12:` for `ldr`)
- `R_AARCH64_AUTH_GOT_ADD_LO12_NC` (`:got_auth_lo12:` for `add`)

`LOADgotAUTH` pseudo-instruction is introduced which is later expanded to actual instruction sequence like the following.

```
adrp x16, :got_auth:sym
add x16, x16, :got_auth_lo12:sym
ldr x0, [x16]
autia x0, x16
```

If a resign is requested, like below, `LOADgotPAC` pseudo is used, and GOT load is lowered similarly to `LOADgotAUTH`.

```
@var = global i32 0
define ptr @resign_globalvar() {
  ret ptr ptrauth (ptr @var, i32 3, i64 43)
}
```

Both SelectionDAG and GlobalISel are suppported. For FastISel, we fall back to SelectionDAG.

Tests starting with 'ptrauth-' have corresponding variants w/o this prefix.

See also specification https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#appendix-signed-got

### CODE REVIEW

**Comment by kovdan01** [comment]
Would be glad to see feedback on the changes from those who are interested.

**Comment by kovdan01** [comment]
Would be glad to see feedback on the changes from those who are interested.

**Comment by kovdan01** [comment]
Ping: would be glad to see feedback on the changes from those who are interested.

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp
Comment (tmatheson-arm) [inline llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp:86]: Please add a brief doc comment

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp
Comment (tmatheson-arm) [inline llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp:96]:
```suggestion
  const auto *PAP = mdconst::extract_or_null<ConstantInt>(
          M->getModuleFlag("aarch64-elf-pauthabi-platform"));
  if (!PAP || PAP->getZExtValue() != ELF::AARCH64_PAUTH_PLATFORM_LLVM_LINUX)
      return false;
```
This avoids the magic -1 and makes the logic easier to follow, imho.

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp
Comment (tmatheson-arm) [inline llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp:102]: The function seems to return `true` if there is no `aarch64-elf-pauthabi-version` flag, which doesn't seem right?

```suggestion
  const auto *PAV = mdconst::extract_or_null<ConstantInt>(
          M->getModuleFlag("aarch64-elf-pauthabi-version"));
  if (!PAV)
      return false;
  return PAV->getZExtValue() &
          (1 << ELF::AARCH64_PAUTH_PLATFORM_LLVM_LINUX_VERSION_GOT);
```

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp
Comment (tmatheson-arm) [inline llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp:33]: We should probably let clang-format format this, does it make a total mess or is the diff just large?

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.h
Comment (tmatheson-arm) [inline llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.h:27]: Ditto

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp
Comment (kovdan01) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp:33]: Allowing clang-format here leads to inconsistent formatting between new and old case labels - see a fragment below

```
  case VK_TLSDESC_PAGE:        return ":tlsdesc:";
  case VK_SECREL_LO12:         return ":secrel_lo12:";
  case VK_SECREL_HI12:         return ":secrel_hi12:";
  case VK_GOT_AUTH:
    return ":got_auth:";
  case VK_GOT_AUTH_PAGE:
    return ":got_auth:";
  case VK_GOT_AUTH_LO12:
    return ":got_auth_lo12:";
```

To make things consistent, I can format old cases the way clang-format likes (with return on new line). This should probably be done as a separate patch in order not to mess functionality and refactoring changes - if so, it's probably OK to leave `// clang-format off` until the follow-up refactoring patch merges. Please let me know your thoughts on this.

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.h
Comment (kovdan01) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.h:27]: Allowing clang-format here breaks custom formatting which IMHO is very nice and helps understanding.

Without clang-format:

```
    VK_CALL              = VK_ABS,
    VK_ABS_PAGE          = VK_ABS      | VK_PAGE,
    VK_ABS_PAGE_NC       = VK_ABS      | VK_PAGE    | VK_NC,
    VK_ABS_G3            = VK_ABS      | VK_G3,
    VK_ABS_G2            = VK_ABS      | VK_G2,
    VK_ABS_G2_S          = VK_SABS     | VK_G2,
    VK_ABS_G2_NC         = VK_ABS      | VK_G2      | VK_NC,
    VK_ABS_G1            = VK_ABS      | VK_G1,
    VK_ABS_G1_S          = VK_SABS     | VK_G1,
    VK_ABS_G1_NC         = VK_ABS      | VK_G1      | VK_NC,
```

With clang-format:

```
    VK_CALL = VK_ABS,
    VK_ABS_PAGE = VK_ABS | VK_PAGE,
    VK_ABS_PAGE_NC = VK_ABS | VK_PAGE | VK_NC,
    VK_ABS_G3 = VK_ABS | VK_G3,
    VK_ABS_G2 = VK_ABS | VK_G2,
    VK_ABS_G2_S = VK_SABS | VK_G2,
    VK_ABS_G2_NC = VK_ABS | VK_G2 | VK_NC,
    VK_ABS_G1 = VK_ABS | VK_G1,
    VK_ABS_G1_S = VK_SABS | VK_G1,
    VK_ABS_G1_NC = VK_ABS | VK_G1 | VK_NC,
```

Can we leave `// clang-format off` here? Or is there probably a better way to teach clang-format how to keep such custom formatting?

File: llvm/test/MC/AArch64/ilp32-diagnostics.s
Comment (MaskRay) [inline llvm/test/MC/AArch64/ilp32-diagnostics.s:88]: Testing caret `^` is not really useful... To test diagnostics, just check the file and line information before `error:`

I'll adjust other diagnostics in the file.

Comment (kovdan01) [reply llvm/test/MC/AArch64/ilp32-diagnostics.s:88]: Yes, I just followed the same pattern as other test cases in the file. I agree that this is not useful w/o `--strict-whitespace --match-full-lines`. I'll delete this, thanks.

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp
Comment (kovdan01) [reply llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp]: I agree, it's much better, thanks! Applied your suggestion in 4c85d6c31cfc89e8543d4f8d78c9cd82dd33ca37

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp
Comment (kovdan01) [reply llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp]: You are right, that's not correct, thanks for pointing this out. This magic with -1 was OK in AsmPrinter (probably should be changes to `std::optional` or smth there as well though), but here it's not applicable. Applied your suggestion in 4c85d6c31cfc89e8543d4f8d78c9cd82dd33ca37

File: llvm/test/MC/AArch64/ilp32-diagnostics.s
Comment (kovdan01) [reply llvm/test/MC/AArch64/ilp32-diagnostics.s]: Fixed in merge commit 83cd3ca82ba0f31bbedf16112236cffb26a3a73b

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp
Comment (kovdan01) [reply llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp:92]: Added, see 76ca5c45a02d2d35ab54ae562cdab0f70ecd14e3, thanks for suggestion

**Comment by kovdan01** [comment]
Ping: would be glad to see feedback on the changes from those who are interested.

Tagging @tmatheson-arm @MaskRay

**Comment by kovdan01** [comment]
Ping: would be glad to see feedback on the changes from those who are interested.

Tagging @tmatheson-arm @MaskRay

**Comment by kovdan01** [comment]
Ping: would be glad to see feedback on the changes from those who are interested.

Tagging @tmatheson-arm @MaskRay

**Comment by kovdan01** [comment]
Ping: would be glad to see feedback on the changes from those who are interested.

Tagging @tmatheson-arm @MaskRay

**Review by tommat01 — APPROVED** [review]

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.h
Comment (tmatheson-arm) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.h:27]: The main advantage of clang-format is to avoid discussing such things 😄

File: llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp
Comment (tmatheson-arm) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp:33]: I would just let clang-format do its thing. Depending on how bad the diff was, I might pre-commit it before the review. Doing it later is fine too though.

Comment (kovdan01) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp:33]: OK, let's allow clang-format to do its job after the PR is merged. I'd prefer not to include large styling changes on unrelated code in this commit and do them separately to distinguish new functionality and style changes. It looks like it would be a bit easier to just delete these `// clang-format off` comments after merging the PR rather than generating a diff, pre-committing it now, resolving merge conflicts...

The same applies to the comment in AArch64MCExpr.h below.

Please let me know if you have some concerns with this.

Comment (tmatheson-arm) [reply llvm/lib/Target/AArch64/MCTargetDesc/AArch64MCExpr.cpp:33]: Sounds good

File: llvm/test/CodeGen/AArch64/basic-pic-auth.ll
Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/basic-pic-auth.ll:21]: Ah, yes, they should, thanks for bringing attention to this.

The tests were initially copied from non-auth variants, and those did not have -NEXT. I'll submit a separate commit adding -NEXT where appropriate in non-auth tests.

File: llvm/test/CodeGen/AArch64/basic-pic-auth.ll
Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/basic-pic-auth.ll:5]: Yes, thanks for suggestion!

Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/basic-pic-auth.ll]: Fixed in 10860a08210f85a35a9b935dbd7aec31dea204e2

File: llvm/test/CodeGen/AArch64/basic-pic-auth.ll
Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/basic-pic-auth.ll]: Fixed in 10860a08210f85a35a9b935dbd7aec31dea204e2

**Comment by kovdan01** [comment]
JFI: merging of this PR is blocked by merging of #96159, which is also already approved, but it's merge is blocked by #96478 :) Hopefully someone submits their feedback - feel free to add reviewers in #96478 if you know someone potentially interested and not added already.

File: llvm/test/CodeGen/AArch64/basic-pic-auth.ll
Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/basic-pic-auth.ll]: I've submitted #101968 enhancing corresponding non-pauth tests

File: llvm/test/CodeGen/AArch64/basic-pic-auth.ll
Comment (MaskRay) [inline llvm/test/CodeGen/AArch64/basic-pic-auth.ll:1]: A filename that starts with `ptrauth-` might be better to group ptrauth tests together.
I am not sure the test should focus specifically on PIC. It's common to test both non-pic and pic in one test.

**Comment by kovdan01** [comment]
@MaskRay Please let me know if you want to leave some additional feedback or if this can be merged. Thanks!

File: llvm/test/CodeGen/AArch64/ptrauth-elf-globals-pic.ll
Comment (MaskRay) [inline llvm/test/CodeGen/AArch64/ptrauth-elf-globals-pic.ll:1]: It might not be necessary to have basic-pic elf-globals-pic.ll extern-weak.ll. You can combine them into one file.

**Review by MaskRay — APPROVED** [review]

File: llvm/test/CodeGen/AArch64/ptrauth-elf-globals-pic.ll
Comment (kovdan01) [reply llvm/test/CodeGen/AArch64/ptrauth-elf-globals-pic.ll:1]: I'll try (as a separate follow-up patch) to combine such tests in two files: one for ptrauth-*.ll tests and one for corresponding non-pauth tests. Thanks for suggestion.

**Pull request merged**
